### PR TITLE
refactor[l2geth]: queue origin type

### PR DESCRIPTION
**Pull request by karlfloersch**
**Description**
Quick refactor fixing the type of `queueOrigin` to no longer be a `big.int` and instead use its type `QueueOrigin`

### CODE REVIEW

File: l2geth/core/types/transaction_meta.go [hunk 1]
```diff
@@ -160,14 +160,10 @@ func TxMetaEncode(meta *TransactionMeta) []byte {
 		common.WriteVarBytes(b, 0, l.Bytes())
 	}
 
-	queueOrigin := meta.QueueOrigin
-	if queueOrigin == nil {
```
Comment (tynes) [inline l2geth/core/types/transaction_meta.go:164]: The queue origin should never be nil anywhere but this is technically a breaking change to the db. I'd like to upgrade my replica to this code to be sure that it just works

Comment (karlfloersch) [reply l2geth/core/types/transaction_meta.go:164]: We really need to add a replica test that we can trigger remotely

File: l2geth/core/state_transition_ovm.go [hunk 2]
```diff
@@ -74,7 +74,7 @@ func AsOvmMessage(tx *types.Transaction, signer types.Signer, decompressor commo
 	// correct format when deserialized from the EVM events, see
 	// rollup/sync_service.go.
 	qo := msg.QueueOrigin()
-	if qo != nil && qo.Uint64() == uint64(types.QueueOriginL1ToL2) {
+	if qo == types.QueueOriginL1ToL2 {
```
Comment (tynes) [inline l2geth/core/state_transition_ovm.go:77]: I wouldn't be opposed to turning 76 and 77 into a single line

File: l2geth/internal/ethapi/api.go [hunk 3]
```diff
@@ -1358,13 +1358,11 @@ func newRPCTransaction(tx *types.Transaction, blockHash common.Hash, blockNumber
...
 			result.L1BlockNumber = (*hexutil.Big)(meta.L1BlockNumber)
 		}
-		if meta.QueueOrigin != nil {
-			switch meta.QueueOrigin.Uint64() {
-			case uint64(types.QueueOriginSequencer):
-				result.QueueOrigin = "sequencer"
-			case uint64(types.QueueOriginL1ToL2):
-				result.QueueOrigin = "l1"
-			}
+		switch meta.QueueOrigin {
```
Comment (tynes) [inline l2geth/internal/ethapi/api.go:1361]: I wouldn't be opposed to making `types.QueueOrigin` implement `Stringer` like so:

Where the type alias is created:
```golang
func (q *QueueOrigin) String() {
  switch q:
  case QueueOriginSequencer:
    return "sequencer"
  case QueueOriginL1ToL2:
    return "l1"
  default:
    return ""
}
```

Then here we can just do:

```
result.QueueOrigin = meta.QueueOrigin.String()
```

File: l2geth/rollup/sync_service.go [hunk 4]
```diff
@@ -676,7 +676,7 @@ func (s *SyncService) applyTransactionToTip(tx *types.Transaction) error {
 		bn := tx.L1BlockNumber()
 		s.SetLatestL1Timestamp(ts)
 		s.SetLatestL1BlockNumber(bn.Uint64())
-		log.Debug("Updating OVM context based on new transaction", "timestamp", ts, "blocknumber", bn.Uint64(), "queue-origin", tx.QueueOrigin().Uint64())
+		log.Debug("Updating OVM context based on new transaction", "timestamp", ts, "blocknumber", bn.Uint64(), "queue-origin", tx.QueueOrigin())
```
Comment (tynes) [inline l2geth/rollup/sync_service.go:679]: If we implemented `Stringer`, then we could do `tx.QueueOrigin().String()` here and have a nice human readable string

File: l2geth/rollup/sync_service.go [hunk 5]
```diff
@@ -782,11 +782,8 @@ func (s *SyncService) ValidateAndApplySequencerTransaction(tx *types.Transaction
 	log.Trace("Sequencer transaction validation", "hash", tx.Hash().Hex())
 
 	qo := tx.QueueOrigin()
-	if qo == nil {
-		return errors.New("invalid transaction with no queue origin")
-	}
-	if qo.Uint64() != uint64(types.QueueOriginSequencer) {
-		return fmt.Errorf("invalid transaction with queue origin %d", qo.Uint64())
+	if qo != types.QueueOriginSequencer {
+		return fmt.Errorf("invalid transaction with queue origin %d", qo)
```
Comment (tynes) [inline l2geth/rollup/sync_service.go:786]: Same for `Stringer`

File: l2geth/rollup/sync_service.go (hunk 4 above)
Comment (karlfloersch) [reply l2geth/rollup/sync_service.go:679]: So I added the `Stringer` and I'm getting:

```
tx.QueueOrigin().String undefined (type types.QueueOrigin has no field or method String)
```

Should this just automatically be converted to the string based on the `Stringer`?

File: l2geth/internal/ethapi/api.go [hunk 6]
```diff
@@ -1358,14 +1358,8 @@ func newRPCTransaction(tx *types.Transaction, blockHash common.Hash, blockNumber
...
-		if meta.QueueOrigin != nil {
-			switch meta.QueueOrigin.Uint64() {
-			case uint64(types.QueueOriginSequencer):
-				result.QueueOrigin = "sequencer"
-			case uint64(types.QueueOriginL1ToL2):
-				result.QueueOrigin = "l1"
-			}
-		}
+
+		result.QueueOrigin = fmt.Sprint(meta.QueueOrigin)
```
Comment (tynes) [inline l2geth/internal/ethapi/api.go:1362]: Did `meta.QueueOrigin.String()` not work?

File: l2geth/core/types/transaction_meta.go (hunk 1 above)
Comment (tynes) [reply l2geth/core/types/transaction_meta.go:164]: I used this code on top of a node that was synced with the old code and confirmed this is backwards compatible under the reasonable assumption that the queue origin is always defined

File: l2geth/core/types/transaction_meta.go [hunk 7]
```diff
@@ -20,14 +20,25 @@ const (
 	QueueOriginL1ToL2    QueueOrigin = 1
 )
 
+func (q QueueOrigin) String() string {
```
Comment (tynes) [inline l2geth/core/types/transaction_meta.go:23]: One last change: 
turning:
```golang
type QueueOrigin int64
```

into 

```golang
type QueueOrigin uint8
```

File: l2geth/internal/ethapi/api.go (hunk 6 above)
Comment (karlfloersch) [reply l2geth/internal/ethapi/api.go:1362]: It was causing an error in my linter saying that String() does not exist

File: l2geth/core/types/transaction_meta.go (hunk 7 above)
Comment (karlfloersch) [reply l2geth/core/types/transaction_meta.go:23]: Done!

**Review by gakonst — APPROVED** [review]

**Comment by tynes** [comment]
I've updated the base of this branch to target the regenesis for safety purposes since it changes the db format

**Pull request merged**